snitun/server/run.py

```diff
@@ -306,50 +310,61 @@ def _process(
         con: socket.socket,
         workers_lb: Iterable[ServerWorker],
         data: bytes,
+        connection_cleanup: ConnectionCleanupHandlerType,
     ) -> None | bytes:
         """Process connection & helo."""
         try:
             data += con.recv(MAX_READ_SIZE)
         except OSError as err:
             _LOGGER.warning("Receive fails: %s", err)
-            self._close_socket(con, shutdown=False)
-            return None
+            return connection_cleanup(con.fileno(), shutdown=False)
 
         # No data received
         if not data:
-            self._close_socket(con)
-            return None
+            return connection_cleanup(con.fileno())
 
         # Peer connection
         if data.startswith(b"gA"):
+            self._poller.unregister(con.fileno())
             next(workers_lb).handover_connection(con, data)
             _LOGGER.debug("Handover new peer connection: %s", data)
-            return None
+            return connection_cleanup(
```

Comment:
let's still return None and call that above. If we ever change `connection_cleanup` which even not have a explicit return None, we have hard time to read the code. It will be simpler to read